mater/lib/src/stores/filestore.rs

```diff
@@ -90,6 +108,86 @@ where
         }
     }
 
+    let index_offset = position;
+    let single_width_index =
+        SingleWidthIndex::new(Sha256::output_size() as u32, entries.len() as u64, entries);
+    let index = Index::MultihashIndexSorted(MultihashIndexSorted::from_single_width(
+        SHA_256_CODE,
+        single_width_index.into(),
+    ));
+    writer.write_index(&index).await?;
+
+    writer.get_inner_mut().rewind().await?;
+    let header = CarV2Header::new(
+        false,
+        car_v1_start.try_into().unwrap(),
+        (index_offset - car_v1_start).try_into().unwrap(),
+        index_offset.try_into().unwrap(),
+    );
+    writer.write_header(&header).await?;
+
+    let header_v1 = CarV1Header::new(vec![root.unwrap()]);
+    writer.write_v1_header(&header_v1).await?;
+
+    writer.finish().await?;
+
+    Ok(root.unwrap())
+}
+
+async fn balanced_import_unixfs<Src, Out>(
+    mut source: Src,
+    mut output: Out,
+    chunk_size: usize,
+    tree_width: usize,
+) -> Result<Cid, Error>
+where
+    Src: AsyncRead + Unpin + Send + 'static,
+    Out: AsyncWrite + AsyncSeek + Unpin + Send + 'static,
+{
+    let chunker = async_stream::try_stream! {
+        let mut buf = BytesMut::with_capacity(chunk_size);
+        loop {
+            let read_bytes = source.read_buf(&mut buf).await?;
+            while buf.len() >= chunk_size {
+                let chunk = buf.split_to(chunk_size);
+                yield chunk.freeze();
+            }
+
+            if read_bytes == 0 && !buf.is_empty() {
+                let chunk = buf.split();
+                yield chunk.freeze();
+                break;
+            } else if read_bytes == 0 {
+                break;
+            }
+        }
+    };
+
```

Comment:
If we're reusing the chunkers across two methods, it deserves to be a separate method.
This logic is complex enough not to be duplicated by copy pastying.